File: 백준/Silver/11722. 가장 긴 감소하는 부분 수열/가장 긴 감소하는 부분 수열.py
import sys

N= int(sys.stdin.readline())

Nlist = [0]+list(map(int,sys.stdin.readline().split()))

dp = [0] * (N+1)

for i in range(1,N+1) :
    
    maxnum = 0 
    for j in range(0,i) :
        if Nlist[j]> Nlist[i] :
            maxnum= max(maxnum,dp[j])

    dp[i] = maxnum+1


print(max(dp))

File: 백준/Silver/11722. 가장 긴 감소하는 부분 수열/가장 긴 감소하는 부분 수열.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        String [] oneline =br.readLine().split(" ");
        int[] narr = new int[N];

        for(int i = 0 ; i<N ; i++)
            {
                narr[i] = Integer.parseInt(oneline[i]);
            }

        int [] dp = new int[N];

        for(int i = 0 ; i<N ; i++)
            {
                for(int j = 0 ; j<i ; j++)
                    {
                        if(narr[i]<narr[j])
                        {
                            dp[i] = Math.max(dp[j]+1,dp[i]);
                        }
                    }
            }

        int answer = 0;

        for(int i = 0 ; i<N ; i++)
            {
                if(dp[i] > answer)
                {
                    answer=dp[i];
                }
            }

        bw.write(String.valueOf(answer+1));
        bw.close();
        
    }
}